samza-core/src/main/java/org/apache/samza/metrics/reporter/MetricsSnapshotReporterFactory.java

```diff
@@ -0,0 +1,140 @@
+    String jobName = getJobName(config);
+    String jobId = getJobId(config);
+    Serde<MetricsSnapshot> serde = getSerde(reporterName, config);
+    Optional<String> blacklist = getBlacklist(reporterName, config);
+
+    MetricsSnapshotReporter reporter =
+        new MetricsSnapshotReporter(producer, systemStream, reportingInterval, jobName, jobId, containerName,
+            Util.getTaskClassVersion(config), Util.getSamzaVersion(), Util.getLocalHost().getHostName(), serde,
+            blacklist, SystemClock.instance());
+    reporter.register(this.getClass().getSimpleName(), registry);
+    return reporter;
+  }
+
+  protected SystemProducer getProducer(String reporterName, Config config, MetricsRegistryMap registry) {
+    SystemConfig systemConfig = new SystemConfig(config);
+    String systemName = getSystemStream(reporterName, config).getSystem();
+    String systemFactoryClassName = systemConfig.getSystemFactory(systemName)
+        .orElseThrow(() -> new SamzaException(
+            String.format("Trying to fetch system factory for system %s, which isn't defined in config.", systemName)));
+    SystemFactory systemFactory = ReflectionUtil.getObj(systemFactoryClassName, SystemFactory.class);
+    LOG.info("Got system factory {}.", systemFactory);
+    SystemProducer producer = systemFactory.getProducer(systemName, config, registry);
+    LOG.info("Got producer {}.", producer);
+    return producer;
+  }
+
+  protected SystemStream getSystemStream(String reporterName, Config config) {
+    MetricsConfig metricsConfig = new MetricsConfig(config);
+    String metricsSystemStreamName = metricsConfig.getMetricsSnapshotReporterStream(reporterName)
+        .orElseThrow(() -> new SamzaException("No metrics stream defined in config."));
+    SystemStream systemStream = StreamUtil.getSystemStreamFromNames(metricsSystemStreamName);
+    LOG.info("Got system stream {}.", systemStream);
+    return systemStream;
+  }
+
+  protected Serde<MetricsSnapshot> getSerde(String reporterName, Config config) {
+    StreamConfig streamConfig = new StreamConfig(config);
+    SystemConfig systemConfig = new SystemConfig(config);
+    SystemStream systemStream = getSystemStream(reporterName, config);
+
+    Optional<String> streamSerdeName = streamConfig.getStreamMsgSerde(systemStream);
+    Optional<String> systemSerdeName = systemConfig.getSystemMsgSerde(systemStream.getSystem());
+    String serdeName = streamSerdeName.orElse(systemSerdeName.orElse(null));
+    SerializerConfig serializerConfig = new SerializerConfig(config);
+    Serde<MetricsSnapshot> serde;
+    if (serdeName != null) {
+      Optional<String> serdeFactoryClass = serializerConfig.getSerdeFactoryClass(serdeName);
+      if (serdeFactoryClass.isPresent()) {
+        SerdeFactory<MetricsSnapshot> serdeFactory = ReflectionUtil.getObj(serdeFactoryClass.get(), SerdeFactory.class);
+        serde = serdeFactory.getSerde(serdeName, config);
+      } else {
+        serde = null;
```

Comment:
I was sticking with existing legacy behavior here. It does seem like it could be nice to use `new MetricsSnapshotSerdeV2()`, but maybe there is a legacy case where no serde is desired, and this is a way to achieve no serde.